inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/AsyncClassTransformer.java

```diff
@@ -211,11 +208,40 @@ private void removeInstrumentationForBatch(List<Class<?>> batchClasses) {
         }
     }
 
+    private DynamicType.Builder<?> instrumentByteCode(Class<?> classBeingRedefined, TypeDescription type, ClassInstrumentationConfiguration classConf, DynamicType.Builder<?> builder) {
+        // Apply the instrumentation hook
+        ElementMatcher.Junction<MethodDescription> hookedMethodsMatcher = getCombinedMethodMatcher(classBeingRedefined, classConf);
+        ElementMatcher.Junction<MethodDescription> contextRequiringMethodsMatcher = hookedMethodsMatcher;
+
+        //find out which methods require a context setup
+        for (SpecialSensor specialSensor : classConf.getActiveSpecialSensors()) {
+            contextRequiringMethodsMatcher = contextRequiringMethodsMatcher.or(specialSensor.getMethodsRequiringContext());
+        }
+
+        //Apply the actual instrumentation onto the builders
+        for (SpecialSensor specialSensor : classConf.getActiveSpecialSensors()) {
+            builder = specialSensor.instrumentPreContextSetup(classBeingRedefined, type, classConf.getActiveConfiguration(), builder);
+        }
+
+        builder = contextAdviceBuilder.applyOuterAdvice(contextRequiringMethodsMatcher, builder);
+
+        //Apply the actual instrumentation onto the builders
+        for (SpecialSensor specialSensor : classConf.getActiveSpecialSensors()) {
+            builder = specialSensor.instrumentPostContextSetup(classBeingRedefined, type, classConf.getActiveConfiguration(), builder);
+        }
+
```

Comment:
I am not a bytebuddy expert, so it's quite hard for me to understand this.. let's explain a bit more with inline comments what is going on here.. also you need to explain me this in details..